runtime/js/40_testing.js

```diff
@@ -823,37 +844,227 @@
     }
   }
 
-  async function runBench(bench) {
-    if (bench.ignore) {
-      return "ignored";
+  function compareMeasurements(a, b) {
+    if (a > b) return 1;
+    if (a < b) return -1;
+
+    return 0;
+  }
+
+  function benchStats(n, t, avg, min, max, all) {
+    return {
+      n,
+      min,
+      max,
+      p75: all[MathCeil(n * (75 / 100)) - 1],
+      p99: all[MathCeil(n * (99 / 100)) - 1],
+      avg: !t ? (avg / n) : MathCeil(avg / n),
+      p995: all[MathCeil(n * (99.5 / 100)) - 1],
+      p999: all[MathCeil(n * (99.9 / 100)) - 1],
+    };
+  }
+
+  function benchMeasureSync(t, fn, step) {
+    let n = 0;
+    let avg = 0;
+    let wavg = 0;
+    const all = [];
+    let min = Infinity;
+    let max = -Infinity;
+
+    warmup: {
+      let c = 0;
+      step.warmup = true;
+      let iterations = 20;
+      let budget = 10 * 1e6;
+
+      while (0 < budget || 0 < iterations--) {
+        const t1 = benchNow();
+
+        fn();
+        const t2 = benchNow() - t1;
+        if (0 > t2) {
```

Comment:
Why is this additional logic necessary? When would the elapsed duration be less than 0? (maybe I'm just misreading this because it's the end of the day and the constant is on the left instead of the right---we should change that)

Reply:
Using perf.now() can lead to negative numbers, but we use bench time op here so it shouldn't be the case.

Reply:
Would you be able to add a comment explaining why along with comments answering the other questions? Thanks!

Reply:
https://doc.rust-lang.org/std/time/struct.Instant.html#monotonicity

another issue is losing precision when converting nanoseconds from u128 to f64